src/Microsoft.TemplateEngine.Cli/CommandParsing/NewCommandInputCli2.cs

```diff
@@ -0,0 +1,305 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.DotNet.Cli.CommandLine;
+using Microsoft.TemplateEngine.Abstractions;
+using Microsoft.TemplateEngine.Edge.Template;
+
+namespace Microsoft.TemplateEngine.Cli.CommandParsing
+{
+    public class NewCommandInputCli2 : INewCommandInput
+    {
+        private ParseResult _parseResult;
+        private IReadOnlyList<string> _args;
+        private string _templateNameArg;
+
+        private Command _currentCommand;
+        private Func<Task<int>> _invoke;
+
+        private IReadOnlyDictionary<string, string> _templateParamValues;
+        private IReadOnlyDictionary<string, IList<string>> _templateParamCanonicalToVariantMap;
+
+        // used for parsing args outside the context of a specific template.
+        private readonly Command _noTemplateCommand;
+        private readonly string _commandName;
+
+        public NewCommandInputCli2(string commandName)
+        {
+            _commandName = commandName;
+            _noTemplateCommand = CommandParserSupport.CreateNewCommandWithoutTemplateInfo(_commandName);
+            _currentCommand = _noTemplateCommand;
+        }
+
+        public bool HasParseError
+        {
+            get
+            {
+                return _parseResult.Errors.Count() > 0;
+            }
+        }
+
+        public int Execute(params string[] args)
+        {
+            _args = args;
+            ParseArgs();
+            bool needsReparse = false;
+
+            if (ExtraArgsFileNames != null && ExtraArgsFileNames.Count > 0)
+            {   // add the extra args to the _args and force a reparse
+                List<string> extraArgs = AppExtensions.CreateArgListFromAdditionalFiles(ExtraArgsFileNames);
+                List<string> allArgs = new List<string>(_args);
+                allArgs.AddRange(extraArgs);
+                _args = allArgs;
+                needsReparse = true;
+            }
+
+            IList<string> templateNameList = _parseResult.GetArgumentListAtPath(new[] { _commandName }).ToList();
+            if ((templateNameList.Count > 0) &&
+                !templateNameList[0].StartsWith("-", StringComparison.Ordinal)
+                && (_parseResult.Tokens.Count >= 2)
+                && string.Equals(templateNameList[0], _parseResult.Tokens.ToList()[1], StringComparison.Ordinal))
```

Comment:
Instead of `ToList()[x]` we could `.ElementAt(x)` - that's assuming of course that this is `ICollection<T>` or `IEnumerable<T>` and doesn't support indexing directly